src/node/stage/approved.rs

```diff
@@ -300,115 +296,273 @@ impl Approved {
+                message: msg_bytes,
+                parsec_version: self.consensus_engine.parsec_version(),
+            },
+        )?;
+        let bounce_msg = bounce_msg.to_bytes()?;
+
+        // If the message came from one of our elders then bounce it only to them to avoid message
+        // explosion.
+        let our_elder_sender = sender.filter(|sender| {
+            self.shared_state
+                .sections
+                .our_elders()
+                .any(|p2p_node| p2p_node.peer_addr() == sender)
+        });
+        if let Some(sender) = our_elder_sender {
+            core.send_message_to_target(&sender, bounce_msg)
+        } else {
+            self.send_message_to_our_elders(core, bounce_msg)
+        }
+
+        Ok(())
+    }
+
+    pub fn handle_bounced_untrusted_message(
+        &mut self,
+        core: &mut Core,
+        dst_key: Option<bls::PublicKey>,
+        mut bounced_msg: Message,
+    ) -> Result<()> {
+        let dst_key = if let Some(dst_key) = dst_key {
+            dst_key
+        } else {
+            trace!(
+                "Received BouncedUntrustedMessage({:?}) - missing dst key, discarding",
+                bounced_msg,
+            );
+            return Ok(());
+        };
+
+        if let Err(error) = bounced_msg
+            .src
+            .extend_proof(&dst_key, &self.shared_state.our_history)
+        {
+            trace!(
+                "Received BouncedUntrustedMessage({:?}) - extending proof failed, \
+                 discarding: {:?}",
+                bounced_msg,
+                error,
+            );
+            return Ok(());
+        }
+
+        trace!(
+            "Received BouncedUntrustedMessage({:?}) - resending with extended proof",
+            bounced_msg
+        );
+
+        let msg = MessageWithBytes::new(bounced_msg)?;
+        self.relay_message(core, &msg)
+    }
+
+    pub fn handle_bounced_unknown_message(
+        &mut self,
+        core: &mut Core,
+        sender: P2pNode,
+        bounced_msg_bytes: Bytes,
+        sender_parsec_version: u64,
+    ) {
+        if sender_parsec_version >= self.consensus_engine.parsec_version() {
+            trace!(
+                "Received BouncedUnknownMessage({:?}) from {} \
+                 - peer is up to date or ahead of us, discarding",
+                MessageHash::from_bytes(&bounced_msg_bytes),
+                sender
+            );
+        }
+
+        trace!(
+            "Received BouncedUnknownMessage({:?}) from {} \
+             - peer is lagging behind, resending with parsec gossip",
+            MessageHash::from_bytes(&bounced_msg_bytes),
+            sender,
+        );
+
+        // First send parsec gossip to update the peer, the resend the message itself. If the
+        // messages arrive in the same order they were sent, the gossip should update the peer so
+        // they will then be able to handle the resent message. If not, the peer will bounce the
+        // message again.
```

Comment:
if bounce again, could that create an infinite loop?

Reply:
No, because eventually their `parsec_version` would catch up and at that point the elder would no re-send the message again. Unless the node is malicious and is purposefully sending `BouncedUnknownMessage` with the same parsec version over and over. But that is a malicious behaviour that can be also caused by other means and probably needs a more general way of handling.

Reply:
ok